src/engine/routing_algorithms/alternative_path_mld.cpp

```diff
@@ -0,0 +1,936 @@
+                                                        alternative_paths_last);
+        }
+    }
+
+    number_of_filtered_alternative_paths = std::min(
+        static_cast<std::size_t>(max_number_of_alternatives_to_unpack),
+        static_cast<std::size_t>(alternative_paths_last - (begin(weighted_packed_paths) + 1)));
+
+    // ^ refactor
+
+    const auto paths_first = begin(weighted_packed_paths);
+    const auto paths_last = begin(weighted_packed_paths) + 1 + number_of_filtered_alternative_paths;
+    const auto number_of_packed_paths = paths_last - paths_first;
+
+    // Todo: benchmark how many packed paths we can unnpack in our time budget.
+
+    std::vector<WeightedViaNodeUnpackedPath> unpacked_paths;
+    unpacked_paths.reserve(number_of_packed_paths);
+
+    // Note: re-uses (read: destroys) heaps; we don't need them from here on anyway.
+    unpackPackedPaths(paths_first,
+                      paths_last,
+                      std::back_inserter(unpacked_paths),
+                      search_engine_data,
+                      facade,
+                      phantom_node_pair);
+
+    //
+    // Filter and rank a second time. This time instead of being fast and doing
+    // heuristics on the packed path only we now have the detailed unpacked path.
+    //
+
+    auto unpacked_paths_last = end(unpacked_paths);
+
+    for (std::size_t i = 1; i < unpacked_paths.size(); ++i)
+    {
+        for (std::size_t j = 0; j < i; ++j)
+        {
+            unpacked_paths_last = filterUnpackedPathsBySharing(
+                unpacked_paths[i], begin(unpacked_paths) + i, unpacked_paths_last);
+        }
+    }
+
+    // Todo: rank a last time by weight and sharing here, then just take first n
+
+    const auto unpacked_paths_first = begin(unpacked_paths);
+    const auto number_of_unpacked_paths =
+        std::min(static_cast<std::size_t>(max_number_of_alternatives) + 1,
+                 static_cast<std::size_t>(unpacked_paths_last - unpacked_paths_first));
+
+    unpacked_paths_last = unpacked_paths_first + number_of_unpacked_paths;
+
+    //
+    // Annotate the unpacked path and transform to proper internal route result.
+    //
+
+    std::vector<InternalRouteResult> routes;
+    routes.reserve(number_of_unpacked_paths);
+
+    const auto unpacked_path_to_route = [&](const WeightedViaNodeUnpackedPath &path) {
+        return extractRoute(facade, path.via.weight, phantom_node_pair, path.nodes, path.edges);
+    };
+
+    // Only now that we annotated the alternatives do we have its actual duration. Note: the
+    // weight we used so far abstracts over duration and includes (potentially high) penalties.
+    // Note: see assembleGeometry for distance calculation
+
+    // Shortest Path
+    routes.push_back(unpacked_path_to_route(*unpacked_paths_first));
+
+    for (auto it = std::next(unpacked_paths_first); it != unpacked_paths_last; ++it)
+    {
+        const auto shortest_path_duration = routes.front().duration_without_turns();
```

Comment:
This should be moved outside of the loop, otherwise this will be recalculated on every run, right?